File: src/main/kotlin/cn/bakamc/folia/db/table/FlightEnergy.kt
package cn.bakamc.folia.db.table

import org.ktorm.database.Database
import org.ktorm.entity.Entity
import org.ktorm.entity.sequenceOf
import org.ktorm.schema.Table
import org.ktorm.schema.boolean
import org.ktorm.schema.double
import org.ktorm.schema.varchar

open class FlightEnergies(alias: String?) : Table<FlightEnergy>("flight_energy", alias) {
    companion object : FlightEnergies(null)

    override fun aliased(alias: String): Table<FlightEnergy> = FlightEnergies(alias)

    val uuid = varchar("uuid").primaryKey().references(PlayerInfos) { it.player }

    val energy = double("energy").bindTo { it.energy }

    val enabled = boolean("enabled").bindTo { it.enabled }

    val barVisible = boolean("bar_visible").bindTo { it.barVisible }
}

interface FlightEnergy : Entity<FlightEnergy> {
    companion object : Entity.Factory<FlightEnergy>()

    var player: PlayerInfo

    var energy: Double

    var enabled: Boolean

    var barVisible: Boolean

    val uuid: String get() = player.uuid
}

val Database.flightEnergies get() = this.sequenceOf(FlightEnergies)

File: bakamc-folia/src/main/scala/cn/bakamc/folia/database/table/FlightEnergy.scala
package cn.bakamc.folia.database.table

import org.bukkit.entity.Player
import slick.jdbc.MySQLProfile.api.*

object FlightEnergy {

  def apply(player: Player, energy: Double, enabled: Boolean, barVisible: Boolean): FlightEnergy =
    FlightEnergy(player.getUniqueId.toString, energy, enabled, barVisible)

}

case class FlightEnergy(
  playerUuid: String,
  var energy: Double,
  var enabled: Boolean,
  var barVisible: Boolean
)

class FlightEnergies(tag: Tag) extends Table[FlightEnergy](tag, "flight_energy") {
  def playerUuid = column[String]("player_uuid", O.PrimaryKey, O.Length(36))

  def energy = column[Double]("energy")

  def enabled = column[Boolean]("enabled")

  def barVisible = column[Boolean]("bar_visible")

  def * = (playerUuid, energy, enabled, barVisible).mapTo[FlightEnergy]
}

val flightEnergies = TableQuery[FlightEnergies]
